src/Common/src/System/Security/Cryptography/KeyFormatHelper.cs

```diff
@@ -0,0 +1,620 @@
+            Debug.Assert(privateKey.Length > 0, "privateKey was empty");
+
+            // https://tools.ietf.org/html/rfc5208#section-5
+            //
+            // PrivateKeyInfo ::= SEQUENCE {
+            //   version                   Version,
+            //   privateKeyAlgorithm       PrivateKeyAlgorithmIdentifier,
+            //   privateKey                PrivateKey,
+            //   attributes           [0]  IMPLICIT Attributes OPTIONAL }
+            // 
+            // Version ::= INTEGER
+            // PrivateKeyAlgorithmIdentifier ::= AlgorithmIdentifier
+            // PrivateKey ::= OCTET STRING
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+
+            // PrivateKeyInfo
+            writer.PushSequence();
+
+            // https://tools.ietf.org/html/rfc5208#section-5 says the current version is 0.
+            writer.WriteInteger(0);
+
+            // PKI.Algorithm (AlgorithmIdentifier)
+            WriteEncodedSpan(writer, algorithmIdentifier);
+            
+            // PKI.privateKey
+            writer.WriteOctetString(privateKey);
+
+            // We don't currently accept attributes, so... done.
+            writer.PopSequence();
+            return writer;
+        }
+
+        internal static unsafe AsnWriter WriteEncryptedPkcs8(
+            ReadOnlySpan<char> password,
+            AsnWriter pkcs8Writer,
+            PbeParameters pbeParameters)
+        {
+            System.Text.Encoding encoding = System.Text.Encoding.UTF8;
+            int requiredBytes = encoding.GetByteCount(password);
+            Span<byte> passwordBytes = stackalloc byte[0];
+            byte[] rentedPasswordBytes = Array.Empty<byte>();
+
+            if (requiredBytes > 128)
+            {
+                rentedPasswordBytes = ArrayPool<byte>.Shared.Rent(requiredBytes);
+                passwordBytes = rentedPasswordBytes;
+            }
+            else
+            {
+                passwordBytes = stackalloc byte[requiredBytes];
+            }
+
+            try
+            {
+                fixed (byte* bytePtr = rentedPasswordBytes)
+                {
+                    int written = encoding.GetBytes(password, passwordBytes);
+
+                    if (written != requiredBytes)
+                    {
+                        Debug.Fail("UTF8 encoding length changed between size and convert");
+                        throw new CryptographicException();
+                    }
+
+                    passwordBytes = passwordBytes.Slice(0, written);
+
+                    return WriteEncryptedPkcs8(
+                        password,
+                        passwordBytes,
+                        pkcs8Writer,
+                        pbeParameters);
+                }
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(passwordBytes);
+
+                if (rentedPasswordBytes.Length > 0)
```

Comment:
Do you need this if check?

Reply:
Does ArrayPool special case the empty array to avoid the throw of "this wasn't my array"? Assuming no, yes :)

Reply:
Yes, empty arrays can be returned.  That said, I don't have a strong preference either way.